### PR TITLE
spring-security-hw

### PR DESCRIPTION
**Pull request by Nonzeq**

### CODE REVIEW

File: src/main/java/com/kobylchak/bookstore/controller/AuthenticationController.java [hunk 1]
```diff
@@ -25,4 +25,5 @@ public UserResponseDto register(@RequestBody @Valid UserRegistrationRequestDto r
             throws RegistrationException {
         return registrationService.register(requestDto);
     }
+
```
Comment (okuzan) [inline src/main/java/com/kobylchak/bookstore/controller/AuthenticationController.java]:
```suggestion
```

File: src/main/java/com/kobylchak/bookstore/mapper/UserMapper.java [hunk 2]
```diff
@@ -3,12 +3,28 @@
...
+    @AfterMapping
+    default void setRoleIds(User user, @MappingTarget UserResponseDto responseDto) {
+        Set<Long> rolesIds = user.getRoles().stream()
+                                .map(Role::getId)
+                                .collect(Collectors.toSet());
+        responseDto.setRoles(rolesIds);
+    }
+
     User toModel(UserRegistrationRequestDto userRegistrationRequestDto);
+
```
Comment (okuzan) [inline src/main/java/com/kobylchak/bookstore/mapper/UserMapper.java]:
```suggestion
```

File: src/main/resources/db/changelog/changes/08-insert-user_roles-table.yml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+                  value: 2
+        - insert:
+            tableName: user_roles
+            columns:
+              - column:
+                  name: user_id
+                  value: 8
+              - column:
+                  name: role_id
+                  value: 1
```
Comment (okuzan) [inline src/main/resources/db/changelog/changes/08-insert-user_roles-table.yml]: don't forget to leave last line empty
image

File: src/main/resources/db/changelog/changes/06-insert-roles-into-roles-table.yml [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+            columns:
+              - column:
+                  name: name
+                  value: USER
+        - insert:
+            tableName: roles
+            columns:
+              - column:
+                  name: name
+                  value: ADMIN
```
Comment (okuzan) [inline src/main/resources/db/changelog/changes/06-insert-roles-into-roles-table.yml]:
```suggestion
                  value: ADMIN

```

File: src/main/resources/db/changelog/changes/04-create-role-table.yml [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: name
+                  type: varchar(255)
+                  constraints:
+                    unique: true
+                    nullable: false
```
Comment (okuzan) [inline src/main/resources/db/changelog/changes/04-create-role-table.yml]:
```suggestion
                    nullable: false

```

File: src/main/resources/db/changelog/changes/05-create-user_roles-table.yml [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+      changes:
+        - createTable:
+            tableName: user_roles
+            columns:
+              - column:
+                  name: user_id
+                  type: bigint
+              - column:
+                  name: role_id
+                  type: bigint
```
Comment (okuzan) [inline src/main/resources/db/changelog/changes/05-create-user_roles-table.yml]:
```suggestion
                  type: bigint

```

File: src/main/java/com/kobylchak/bookstore/model/User.java [hunk 7]
```diff
@@ -1,19 +1,28 @@
...
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
 
 @Getter
 @Setter
 @Table(name = "users")
 @Entity
-public class User {
+public class User implements UserDetails {
```
Comment (okuzan) [inline src/main/java/com/kobylchak/bookstore/model/User.java:29]: I believe user should implement soft delete at this point

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/com/kobylchak/bookstore/service/user/RegistrationServiceImpl.java [hunk 8]
```diff
@@ -22,7 +25,16 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
             throw new RegistrationException(
                 "User with email: " + requestDto.getEmail() + " already exist");
         }
+        requestDto.setPassword(passwordEncoder.encode(requestDto.getPassword()));
         User user = userRepository.save(userMapper.toModel(requestDto));
         return userMapper.toDto(user);
     }
+
+    @Override
+    public List<UserResponseDto> getAll() {
```
Comment (Rommelua) [inline src/main/java/com/kobylchak/bookstore/service/user/RegistrationServiceImpl.java]: Why do we need this method? Delete it.

File: src/main/java/com/kobylchak/bookstore/service/user/RegistrationService.java [hunk 9]
```diff
@@ -3,8 +3,11 @@
 import com.kobylchak.bookstore.dto.user.UserRegistrationRequestDto;
 import com.kobylchak.bookstore.dto.user.UserResponseDto;
 import com.kobylchak.bookstore.exception.RegistrationException;
+import java.util.List;
 
 public interface RegistrationService {
 
     UserResponseDto register(UserRegistrationRequestDto requestDto) throws RegistrationException;
+
+    List<UserResponseDto> getAll();
```
Comment (Rommelua) [inline src/main/java/com/kobylchak/bookstore/service/user/RegistrationService.java]: Why do we need this method? Delete it.

File: src/main/java/com/kobylchak/bookstore/dto/user/UserResponseDto.java [hunk 10]
```diff
@@ -9,4 +10,5 @@ public class UserResponseDto {
     private String firstName;
     private String lastName;
     private String shippingAddress;
+    private Set<Long> roles;
```
Comment (Rommelua) [inline src/main/java/com/kobylchak/bookstore/dto/user/UserResponseDto.java]: We do not need this field here

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/com/kobylchak/bookstore/repository/user/UserRepository.java [hunk 11]
```diff
@@ -1,9 +1,15 @@
...
 import java.util.Optional;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
 
 public interface UserRepository extends JpaRepository<User, Long> {
+    @Query("from User u join fetch u.roles where u.email = :email")
     Optional<User> findUserByEmail(String email);
+
+    @Query("from User u join fetch u.roles")
+    List<User> findAllUsers();
```
Comment (Rommelua) [inline src/main/java/com/kobylchak/bookstore/repository/user/UserRepository.java]: This method is not used - delete it

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/com/kobylchak/bookstore/model/User.java [hunk 12]
```diff
@@ -1,19 +1,32 @@
...
+import org.hibernate.annotations.SQLRestriction;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
 
 @Getter
 @Setter
 @Table(name = "users")
+@SQLRestriction("is_delete=false")
+@SQLDelete(sql = "UPDATE users set is_deleted = true WHERE id=?")
```
Comment (MAMentorFX) [inline src/main/java/com/kobylchak/bookstore/model/User.java:27]:
```suggestion
@SQLRestriction("is_delete = FALSE")
@SQLDelete(sql = "UPDATE users SET is_deleted = TRUE WHERE id = ?")
```

File: src/main/java/com/kobylchak/bookstore/repository/user/UserRepository.java [hunk 13]
```diff
@@ -3,7 +3,9 @@
 import com.kobylchak.bookstore.model.User;
 import java.util.Optional;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
 
 public interface UserRepository extends JpaRepository<User, Long> {
+    @Query("from User u join fetch u.roles where u.email = :email")
```
Comment (MAMentorFX) [inline src/main/java/com/kobylchak/bookstore/repository/user/UserRepository.java:9]:
```suggestion
    @Query("FROM User u JOIN FETCH u.roles WHERE u.email = :email")
```

**Review by MAMentorFX — APPROVED** [review]

File: src/main/java/com/kobylchak/bookstore/service/user/RegistrationServiceImpl.java [hunk 14]
```diff
@@ -7,13 +7,15 @@
...
 
 @Service
 @RequiredArgsConstructor
 public class RegistrationServiceImpl implements RegistrationService {
     private final UserRepository userRepository;
     private final UserMapper userMapper;
+    private final PasswordEncoder passwordEncoder;
 
     @Override
     public UserResponseDto register(UserRegistrationRequestDto requestDto)
```
Comment (MAMentorFX) [inline src/main/java/com/kobylchak/bookstore/service/user/RegistrationServiceImpl.java:21]:
```suggestion
@Transactional
    public UserResponseDto register(UserRegistrationRequestDto requestDto)
```

**Pull request merged**